### ISSUE TITLE
IndexError: index out of range in self

### ISSUE THREAD

**Issue by Jacob-yen** [antonio-mastropaolo/robustness-copilot#1]
Hi~ @antonio-mastropaolo  

When using the released tool to rewrite sentences, I encountered an issue. When given the input sentence "Restrictions: 1 <= num <= 1000", the code throws an IndexError exception. The following code is almost identical to the code provided in the file `Code/pre-processing/Pegasus/pegasus.py`. 

```

import torch
import pandas as pd
from transformers import PegasusForConditionalGeneration, PegasusTokenizer
from tqdm import tqdm
from scipy.special import softmax
from scipy.stats import entropy
from nltk.tokenize import word_tokenize
import sys,math

def main():

    model_name = 'tuner007/pegasus_paraphrase'
    torch_device = 'cpu' #'cuda' if torch.cuda.is_available() else 'cpu'
    tokenizer = PegasusTokenizer.from_pretrained(model_name)
    model = PegasusForConditionalGeneration.from_pretrained(
            model_name).to(torch_device)

    sentence = "Restrictions: 1 <= num <= 1000"


    def get_response(input_text, num_return_sequences=None, num_beams=None, temperature=None):
        
        input_ids = tokenizer([input_text], 
                            truncation=True, padding='longest',
                            max_length=100, return_tensors="pt")
       
        generated_outputs = model.generate(**input_ids, max_length=100,  do_sample=False,
                num_beams=1,  return_dict_in_generate=True, output_scores=True)
        
       
        gen_tokens_seq = [tokenizer.decode(g, skip_special_tokens = True) for g in generated_outputs.sequences]
        num_generated_tokens = len(generated_outputs['scores'])
        gen_ids = []
        probs = []
        
        for i1 in range(0, num_generated_tokens, 1):
            temptensor = generated_outputs['scores'][i1][0]
            gen_id = torch.argmax(temptensor).item()
            log_prob = generated_outputs['scores'][i1][0][gen_id].item()
            probs.append(log_prob)
            gen_ids.append(gen_id)
        
        gen_probs = torch.tensor(probs)
        gen_probs = gen_probs.view(1,-1)
        
        unique_prob_per_sequence = torch.softmax(gen_probs, dim=-1)#oftmax(gen_probs)#.sum(-1)/num_generated_tokens #diving over the length to normalize
        entropy_prediction = entropy(unique_prob_per_sequence[0].tolist(), 2)
        
        gen_ids = torch.tensor(gen_ids)
        gen_ids = gen_ids.view(1, -1)
        gen_ids_seq = [tokenizer.decode(g, skip_special_tokens = True) for g in gen_ids]
        print("\nPerturbed: {}\nNormalized Prob: {}\nEntropy: {} \n\n".format(' '.join(gen_ids_seq),unique_prob_per_sequence, entropy_prediction)) 


    perturbed,confidence = get_response(sentence)
    print(perturbed)

main()

```

### Steps to Reproduce
- Use the code provided in the "Code/pre-processing/Pegasus/pegasus.py" file
- Input the sentence "Restrictions: 1 <= num <= 1000"
- Observe the IndexError exception

```
Traceback (most recent call last):
  File "/home/user01/code-models/baselines/pegasus/pegasus.py", line 59, in <module>
    main()
  File "/home/user01/code-models/baselines/pegasus/pegasus.py", line 56, in main
    perturbed,confidence = get_response(sentence)
  File "/home/user01/code-models/baselines/pegasus/pegasus.py", line 28, in get_response
    generated_outputs = model.generate(**input_ids, max_length=100,  do_sample=False,
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/transformers/generation/utils.py", line 1518, in generate
    return self.greedy_search(
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/transformers/generation/utils.py", line 2285, in greedy_search
    outputs = self(
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/transformers/models/pegasus/modeling_pegasus.py", line 1405, in forward
    outputs = self.model(
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/transformers/models/pegasus/modeling_pegasus.py", line 1260, in forward
    decoder_outputs = self.decoder(
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/transformers/models/pegasus/modeling_pegasus.py", line 1035, in forward
    positions = self.embed_positions(input_shape, past_key_values_length)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/transformers/models/pegasus/modeling_pegasus.py", line 140, in forward
    return super().forward(positions)
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/nn/modules/sparse.py", line 160, in forward
    return F.embedding(
  File "/home/user01/.conda/envs/transformers/lib/python3.10/site-packages/torch/nn/functional.py", line 2210, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
IndexError: index out of range in self
```

I would appreciate any help in resolving this issue. Thank you for sharing such wonderful tools.

**Comment by antonio-mastropaolo** [comment]
Hi @Jacob-yen 

First of all, thanks for your appreciation :)

I looked at the code and can confirm that it happens when the model cannot paraphrase the sentence provided as input (e.g., is too short). For example, if you try a different sentence like "This item would be super cool!", no exceptions are raised.

I hope this clarifies your doubt.

**Comment by Jacob-yen** [comment]
> Hi @Jacob-yen
> 
> First of all, thanks for your appreciation :)
> 
> I looked at the code and can confirm that it happens when the model cannot paraphrase the sentence provided as input (e.g., is too short). For example, if you try a different sentence like "This item would be super cool!", no exceptions are raised.
> 
> I hope this clarifies your doubt.

@antonio-mastropaolo 
Thanks for your prompt response! 
I think that the length of the input sentence may not be the root cause of the code exceptions. To test this, I changed the input to a shorter sentence, "I like apples," which did not result in any exceptions. 
Besides, the code works well by changing the input 'Restrictions: 1 <= num <= 1000' to 'Restrictions: 1 num 1000'. It seems that the symbol "<=" was causing the exceptions.
I am curious whether the 'tuner007/pegasus_paraphrase' model can handle code descriptions that contain special symbols such as "<=" and "==." Or is it limited to only paraphrasing sentences made up of common words?

**Comment by antonio-mastropaolo** [comment]
Hi @Jacob-yen 

A quick fix I see to handle these cases, is to build a map and translate these problematic token.
For example, I would translate "<=" into "less or equal" in your case.

It's likely that the Pegasus Paraphrasing model proposed by the authors still needs to be equipped to handle these cases.
Indeed, when building our dataset, we needed to discard instances for which the paraphrasing DL model was raising exceptions. It is also possible that other tokens (e.g, non-ascii) lead the model to these failures.

Anyway, if you want to get to know more, here https://huggingface.co/tuner007/pegasus_paraphrase
the HF repos of the model we used to generate the paraphrased code description.
